### ISSUE TITLE
How to extract user info from session cookie _oauth2_proxy_0 

### ISSUE THREAD

**Issue by hirendave47** [oauth2-proxy/oauth2-proxy#2101]
I am getting below cookie in my frontend app (Angular SPA) after successfully authenticating with Keycloak via oauth2-proxy.

```
_oauth2_proxy_0=2oyJeXaqMhH1LXhdtbal5DfFSNTwR2uT8pIDGaJuOimsHPUzBe5H004kzwZhwBs2Qg86NI6v1GuEZPEfm8jDdptHsVJKR9NbOEsgbNklGhRy_iz9T-q8IQ7Nq7fiXOkTaGZQc7Y3hMf2YET4GFiixWfUng0hEqN4C_I5grU4n2xE2NScLYpvHoKHaHhybnA7TeTPxEukripRI_2kmu4sDOUBqMkzHavCyUHD6EP7c42zqorfPKeVBIcMVH84TK1Yzl8e23Np8CUcbnJ7W-up7ZdDxbwjUO-bU9ivmXQBlhRHp6CX8eTMDAe9ayYBuyr-1qs6c6AcwUYqsVqSU7tTeh5G-F6yJAWbfTA3PqwygRa6PKGydnBgpSRpm61ZrFUCVKskUhjLOah6wOYHBpeVnHB7jXq6nSMq-Am72KS7E3_reAG_s3WXlHyFKB4LfNWV9ER_oTjJRrLZTAkONMHWgJzDYdi8w00G_MqZu8CkXH8UavkEu1yxWkJbB1yBX59-oywC8alffMCtLrsQ9IR9gZL8PIg1fovrUUtPlw67N3OJmkA-35RgIu-QPWQbvG9TKnfcQqBSl5G_K2KLT6Wlfclo-LGfnQBBR_VmxcyXU4tyI-IvlfqyoIlPRvhl-xqvBS58UtZOalRSEaRuW2t92lkYbt99sB4LgOFt64KvxzgzcCBUtqhcbWe8a38P55TWfrrqLrajQ_IUAyYdp-XbLF8mhQElxpwhOE85u9a4nX48Mq9G3IKjbTh4CA2W_B_f_r3gLibQiKK96ldZsPDxsaBr_1OpTbh5s3_DpDDhyEnkHFwtgt9kQArnB03Uka5m6SNIAI2dh5RBhxzK4_d83cfB8EZCceyUhimSf0qAX82DwXcVZOl-wBI33EG0jes4lUDLOPVhO9ntvAsWpLuOWJe0pIanAt1JYIJMPcV41h43yCqoJHXCJiyzxFIFf-lJgzc80lml13zMlb5--8-jUpH1HWZ1JbAKlLnq7fqDsVwwlP3EhhfD_z5ROG8c-iVgG3ZcsXXA-NcsUtvr5KZLXoMJDnj8B4C11GVID9795SeJN0lnLI3gIXCkWcGDLm-lsB0Dal8Huwn9HgA7QhPoN1wbpCCMxj3zjWM-3htLW80ge77KMW8Mj-v5KNsvEtpZ4Y6rrjD0YD5ctu8kJL-SH7vrvtc-0FjPJMz6Qaxos-DjpYAvYXKtytAWdusMKnCZcTm_pV5Iju16X9Q1kykGzIqEogWoipNgg6_OIWY5SlpCNUWvvm6ZE22CHZ5q5aCKggjwYQ-JIf2D5UvCInMqUtzlSbehDadVpib5UIXkA0h7uqX1_BHY9ePjQDphFOBDbHmuWAkrxVZBR-AoBr46qbGoZPNE6xn94i0UlkL9IYv6C62IN0ZlCetHwGuh0nLQ0J3-3wfPfT-J7wboiuPSmbpNWldJTq6HtyXe_t1IQtjg7Sy9TSMDfQl389otbNrnCqbNpBVGbGc_Wul6ttdmjCX8J3lcSrUHbjz4gANwXIuobCt1mghFY0rFqWxx9IwZ8HkNyriXMNke55lsFQI1Op76lOpjzvktIw6VEG0ytcYOU0P3oCgouHHjU4yGN-ZAwN2u1e86cVnpC8K0mqW-xl01dNu-52yKhQBO0N9EVp6RufSEGoY2gzoiIRgCoUvsfxV547XCF_PlVvmsVXt-a0Ho7wQIASvMeEoS9vVYIAePeSIHEgs4XVUlOnW45bpi-fikhOZHHQ-HUh_CSrW2Me3WBeN-EM5NweIQoqs0BI0w1IHOLVBIFGuXx2y0CuJOEa_9ghbV2m7BYFMNMCAVrQma4xkY-XdyJK9bk1BHWMLYlZ2Daw6WOK_FfSbzVKl83NduqCJvu-0DzTp2vO5st15SYKSfH3l4vCL2idRzqTffOMQRFPWxe28aFSfs72r2wgiF0iQI1vXr7TMWNRyBzvbjpEWfcsTNjvIamOGIcLOUQgYshOh9RD6RhLaUUopqnVRc6PEOMdTM7Dexi8cWwqjsV5cHyjDUP_CuucdizRq2ND7OjpuqC9JoY26GFNbRY8lWXN02lU-i0uXk4Dc5DykUsvazevjrDIbFqj8ClW0VzRJI2DEC9XFy3XG2c2r4UZ1vKkW5uo64KGrNJxHiZqJOF8IikJ1Z6xrpdrlpkb_g4gPeEMOfQ1LSueeFwq7inRouyWrPyF1hknkpZEqYYyZ1znVPa3E2InQMefGYrBgwQwPybliYi_9uU1X6h0DpLyYFRMsP1iD1OkPxdKJvZdFaiKsSeERqaO2CURxoPL3LpIVdQPDBS7U40twYhfXOZ036leuLuacyQ5xzzmkk4Qf7iJDgsp0MSwSPtGSzb5UlHhbdjfTOjT6h69idFChsNIyDIP1bRG_D-bRrP82YpDnUW144goTRi-Ek0AZannszT_efk2A36-cGDCeFvmV1-bVrlxCrO33BFySW--fS3NnpgfdesfEML16L4POqsRHyVO-7fn40gLWgsj7U4tPndThAykVcT1HbEtRqRI7_w6aBS41vnUcU-rSY-J3syaevv4zaTWVbhPeyiXQ1qNbyxoOwA-6hOcPjYtvN8UVjY_WYEPExQ25EIUw2n4AvK7QGgLpRstHyZxBaoSOuUcGXLaa6Y0VZS-0vJzT_LD-vSdt1ifCh5O5ygwWgNt8yNuDOLOxnfSCX4Am4F8QW3e3dMrYd5Q4xOh58pkhvYh8168RQlkAeC671wy_zMEisgMOlBH1QLPu0-bH-Df2bLub4qVN9ez6xvTKUA1xc4dyQTkanGhOdQhWp07joGoCgDVi9mw8vJAC6W8Tn0u4bn4Y_U91gTw2OThJiZnaMXDQkYaDiq4dQ3lGUgNG8H7BsdUaL-6kD4SypRTR8I7T0vbIqGGbf3_gVKDzrX7Mj6NmQSKt3mgm2YQHWi1rThZuUtwZ_NPrmKy1hmL3ZDSgkZFyoC1yeTDJqVcx6ZEW_Fnse3FKhDpibmRZaWmspQ3XiWmLtQkBDf06-JQeNIoe4XQiFvwXHju3j_Jx3Nj5qV0FU9jiwoOlSUuBCYNaML4QIjFSFtyNzLF-4yGzZnqP_88YxIcbSBdEdbKSNm2ayf0EYUT7SzPhQpErWvygpw54EuDqmDIXy8xp6oqcVsXutUAKbhirjqes_g3WpnJef4pp0ehJNt2Y-b8qhTrrrkKXsCJ1CgJJTNtRU68tDi0nSjpGEzvEbkorVWqKr8rZAfYjk8P-pPd2gLU-C276zP_CZGqaIKC8cvrdBO3b9-sRVhQ2bEUAdrP2kMqnjsQz0MNEaPyAPwcAlthaStYrzV98kbDgHS8Pmt86pWEZgLypMIZigA8gptRCRNE_T5xkfGrHTA3Fi1_6qDv8E05LqMBThTcAiNxpyWMYHypNfw8MFO3KHRUP2w7Bf8y1tMbruQ5c4OZ-CW388cOqEP6hJJbUrFsN1qhcGL_HQN6auGBIUVNt8Wh4EwKletfNyl23ew8o6rLBHM9eXlLYFMh2vQRp8Hz00fCJs8iy9Zz420CYIMGjNRSTw6ErTMCCuIr8cproIjGfMN2N2sJawu8ZExsb5trPJ7bnTOOm0yCUFc63nGlKdq-KTaYcbRNn0B9AdU3I_OBLA7eFVwXrAhleaNOI37d9Ha7rPHI9R6HUWMGBUVHKj_JoY8zrTpD9pY3EDBOiTm3s-NyRO2Y1GlDE3FP-5oi6YdHd-Xx8oSZ1tWc4WO-N-7uQRqm25LTHMxjnl9wR9yCyhLJ9I3dRSbt9CPXOI8xFLrYVYuvQcxkRCrczuE0lr5vnRmj-wkb9s7JR7q6bQAnDBIJZVUHsVuHoPffW7cgMsuCJSxY1FO5M-P0bXEoCI1UcX_J3uEs3wRiTXVyngAwxye8age6zfODTgZehk6bTRZWdqM02vnFDs2DXW1a2f3lPaMPR5XZiOw-z; _oauth2_proxy_1=Df2aAr0A6qNQE5_zMkFlDN1EvDDjy7nrrZxZ_chT9te4IMkuoS90_UNKZmdk7aztb2zU1y7ZmdNpBUHCP9fi4GMh7Pkif_Q67WXRCs35JnIIU8bf4zM4GA-DO|1682578328|SkLPz356msH6446LVFCJ1tb4iDXB8wpkPaEmzCSWP18=
```

I want help on decoding this cookie to extract information like ID Token, Access Token, Username, Email, Groups etc. and want to pass ID Token to backend for authentication.

Below is my oauth2-proxy configuration...
```
client_id = "urnstack"
client_secret = "tFTxCbb4cziFCfP3Jmu5N35tOQ0gUZlS"
pass_access_token = true
pass_authorization_header = true
pass_user_headers = true
set_xauthrequest = true
set_authorization_header = true
cookie_secret = "mssgFme6-H-HLP-alADLgDMmGQiKLay09Vio5sXlWZg="
oidc_issuer_url = "https://vmserver3.eastus.cloudapp.azure.com:32358/realms/xyz"
provider = "keycloak-oidc"
upstreams = [
     "https://vmserver3.eastus.cloudapp.azure.com:32462"
]
email_domains = "*"
redirect_url = "https://vmserver3.eastus.cloudapp.azure.com:32462/oauth2/callback"
redeem_url = "https://vmserver3.eastus.cloudapp.azure.com:32358/realms/xyz/protocol/openid-connect/token"
request_logging = true
skip_provider_button = true
whitelist_domains = [
   "vmserver3.eastus.cloudapp.azure.com:32462/oauth2/callback",
   "vmserver3.eastus.cloudapp.azure.com:32462",
   "vmserver3.eastus.cloudapp.azure.com",
]
cookie_domains = [
    "vmserver3.eastus.cloudapp.azure.com",
    "vmserver3.eastus.cloudapp.azure.com:32462",
    ".eastus.cloudapp.azure.com",
    "*"
]
code_challenge_method = "plain"
```

Thanks in advance for assistance.

**Comment by hirendave47** [comment]
I've figured out solution for above issue. But I have another doubt now.
If I use redis storage instead of cookie, how does the authentication flow works?
I am using Angular 14 for Frontend development and Golang for backend API development and Keycloak as an identity provider.
Thanks in advance for any assistance.

**Comment by pievalentin** [comment]
Can you provide your solution? I would be interested to implement something similar

**Comment by hirendave47** [comment]
The solution I've implemented is a work around and not secure as I am passing auth token to Frontend directly and from there, sending it to Backend so it can authenticate with middleware.  This is because I haven't figured out yet how to pass token to Backend only and redirect page from oauth2-proxy login to Frontend. The oauth2-proxy passes token to the same URL which is set as redirect URL. If I set redirect URL as frontend and upstream as Backend, I don't know how to fetch token from oauth2-proxy in Golang Backend. If my understanding is correct, implementing secure solution involves passing token to BE instead of FE and for this to work, BE has to connect with Keycloak/oauth2-proxy using access token received from Frontend. It's quite complicated for me so I am using below function to extract required tokens in Frontend directly as of now.

import jwt_decode from 'jwt-decode';

getDecodedAccessToken(token: string): any {
        try {
            return jwt_decode(token);
        } catch (Error) {
            return null;
        }
    }
